### ISSUE TITLE
Reserach Load Test tool

### ISSUE THREAD

**Issue by varokas** [bigbearsio/strength#11]
Ref #10 

**Requirements**
1. We can simulate the entire client flow. Not just calling the same endpoints repeatedly. 
    1. Get available tickets
    2. Book the ticket
    3. Confirm ticket
    4. Repeat until no ticket left
1. We can generate N users (with control on ramp up)
2. Can start work on a distributed machine
3. Each client keeps buying until the tickets are gone
4. Each client stop when the tickets are gone
5. Each client forward sales result to central repo
6. Can detect when client are all done, and then can compute when the tickets are all gone

Some research notes

maybe try to use load test tool like Locust or Gattling because it is asynchronous. 

Benchmark is in locust favor. https://www.blazemeter.com/blog/eight-reasons-you-should-use-gatling-for-your-load-testing

but need to be careful on how to do async-wait

https://github.com/pglass/how-do-i-locust

**Comment by supaket** [comment]
Can we research on Gatling @boatkorachal @tspn

**Comment by varokas** [comment]
yes. but please research on cluster mode for battling as those aren't support out of box

**Comment by supaket** [comment]
Repo here https://github.com/bigbearsio/strength-load

**Comment by supaket** [comment]
@boatkorachal 

> I know another load testing tool name k6 krub. written in Go and can write script to mimic virtual user in javascript and can be plugged to influxdb to store result
> 
> but i have some question. how can we order the request that will be hit the target?
> like if we spawn 6k client try to book tickets. request ordering might different between tests pao krub

**Comment by varokas** [comment]
I'm not sure what it is you meant by request ordering. Here's how I look at it. 

We will spawn 6k people trying to get 100 tickets. Those 6k people will try to secure the ticket via the flow in `strength-ref`, but only 100 of those people will pass. We don't care who of 100 gets it. We don't care which order they gets it.

as soon as a person gets a ticket they will quit. as soon as a person finds out (5900 of them) that a tickets are all gone they will quit

Then we check that server distribute all tickets without duplicates and the seats confirmed in server agrees with what all 100 successful client reports
